capture.js

```diff
@@ -0,0 +1,53 @@
+const capture = {
+
+  init() {
+    this.addListeners();
+    this.prevSite = null;
+    this.nextSite = null;
+  },
+
+  addListeners() {
+    // listen for each HTTP response
+    browser.webRequest.onResponseStarted.addListener(this.collectRequests, {urls: ['<all_urls>']}
+    );
+    // listen for when user visits a new site
+    browser.tabs.onUpdated.addListener(this.checkNewSite);
+  },
+
+  async collectRequests(response) {
+    const tab = await browser.tabs.get(response.tabId);
+    let tabUrl = new URL(tab.url);
+    tabUrl = tabUrl.hostname;
+    let targetUrl = new URL(response.url);
+    targetUrl = targetUrl.hostname;
+    let originUrl = new URL(response.originUrl);
+    originUrl = originUrl.hostname;
+
+    // capture third party requests
+    if (targetUrl !== originUrl) {
+      const thirdPartyData = {
+        document: tabUrl,
+        target: targetUrl,
+        origin: originUrl,
+        requestTime: response.timeStamp
+      };
+      console.log('storage.thirdPartyRequest:', tabUrl, thirdPartyData);
+    }
+  },
+
+  checkNewSite(tabId, changeInfo, tab) {
```

Comment:
Also, looking at my console and clicking around in `about:debugging`, neither it nor about:blank have made any HTTP requests, do some `about:*` pages make requests? Seems like if I don't make calls to the storage API except in cases of HTTP requests/responses, maybe I don't need to filter for these non-visible i.e. about:* tabs?
If I did, what about something like this:
```javascript
if (tab.status === 'complete' && !(re.test(tabUrl))) {
  const firstPartyData = {
    faviconUrl: tab.favIconUrl,
  };
  console.log('storage.firstPartyRequest:', tabUrl.hostname, firstPartyData);
}
```